### PR TITLE
Enhance/slm connectors

### PR DESCRIPTION
**Pull request by Killthebug**
The PR uses a combination of `asyncio` and multithreading to consume the streaming endpoint and yield the chunks in a synchronous manner.

Here's a breakdown of the code:

1. The `chat_completion` method takes various parameters and checks if streaming is enabled. If streaming is enabled, it calls the `_stream_generator_wrapper` method; otherwise, it calls the `_perform_request` method.

2. The `_perform_request` method sends a regular POST request to the specified URL with the provided request data and headers. It returns the response text.

3. The `_stream_generator_wrapper` method is responsible for wrapping the asynchronous streaming generator to be consumed like a synchronous generator. It creates a queue to hold the streamed chunks and starts a separate thread to run the `_consume_streaming_endpoint` method asynchronously.

4. In the `_stream_generator_wrapper`, the code enters a loop that retrieves chunks from the queue until it receives a `None` value, which signals the completion of streaming. It yields each chunk to the caller.

5. The `_consume_streaming_endpoint` method is an asynchronous generator that consumes the streaming endpoint. It uses `httpx.AsyncClient` to send a POST request to the specified URL with the provided request data and headers. It then asynchronously iterates over the response chunks using `aiter_text()`.

6. For each chunk received the code loads the chunk as JSON, and puts it into the queue. After all chunks are processed, it puts `None` into the queue to signal completion.

@nsosio - To support this end-end we need to add a prem-ml production environment key to this prem-utils. Can you please comment if this is the best way to auth prem-ml?

### CODE REVIEW

File: prem_utils/connectors/prem.py [hunk 1]
```diff
@@ -131,28 +57,84 @@ def chat_completion(
         seed: int | None = None,
         stop: str | list[str] = None,
         stream: bool = False,
-        temperature: float = 1,
-        top_p: float = 1,
+        temperature: float = 0.1,
+        top_p: float = 0.9,
```
Comment (nsosio) [inline prem_utils/connectors/prem.py:61]: why these parameters have changed?

File: prem_utils/connectors/prem.py [hunk 2]
```diff
@@ -15,114 +19,36 @@ def __init__(
...
-            },
-        }
         self._api_key = api_key
+        self.prem_ml_url = prem_ml_url
+        self.prem_ml_key = os.environ["PREM_ML_KEY"]  # To authenticate prem_ml slm completion
 
     def parse_chunk(self, chunk):
         return {
-            "id": chunk["id"],
+            "id": chunk.get("id", ""),
```
[the remark this answers is not in the document]
Comment (Killthebug) [reply prem_utils/connectors/prem.py]: Updated

File: prem_utils/connectors/prem.py (hunk 1 above)
Comment (Killthebug) [reply prem_utils/connectors/prem.py]: Updated

File: prem_utils/connectors/prem.py [hunk 3]
```diff
@@ -15,114 +20,36 @@ def __init__(
         self,
         api_key: str,
         base_url: str = "https://ml-development.prem.ninja/api/ml",
+        prem_ml_url: str = "https://ml.premai.io/api/ml/slm-completion/",
```
Comment (nsosio) [inline prem_utils/connectors/prem.py:23]: I've missed this. Both should point to ml.premai.io, there is no dev anymore.  Also, I would only keep the base_url and handle the slm_completion url internally

**Review by nsosio — COMMENTED** [review]
Besides this last comment LGTM

File: prem_utils/connectors/prem.py (hunk 3 above)
Comment (Killthebug) [reply prem_utils/connectors/prem.py]: Updated

**Review by nsosio — APPROVED** [review]

**Pull request merged**